drivers: i2c: stm32: Fix routing of secondary target address

timwoolliscroft-opteran opened the pull request:
Fixes the routing of the events associated with a secondary i2c target address being routed to the primary config.
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/65527

timwoolliscroft-opteran commented:
Apologies @erwango this is duplicate of https://github.com/zephyrproject-rtos/zephyr/pull/65531 - which I have now closed.

Fighting git commit renaming was proving more than my skills could manage and for the 1 line change it seems easier and tidier to resubmit.

erwango reviewed (changes requested):
Thanks for this fix.
Can you elaborate the commit message to better understand the change ?

timwoolliscroft-opteran commented:
> Thanks for this fix. Can you elaborate the commit message to better understand the change ?

Done my best! Effectively its a very simple order of operations bug.

erwango commented:
Change as itself looks fine. But then the if/else section in between should then be modified.
This should be done with https://github.com/zephyrproject-rtos/zephyr/issues/55987 in mind. 
Also I might be wrong but one bug remain about the secondary slave matching in case address is 10bits. This is not related to your change, but if you have the set up to check this, it would be great if you can confirm my doubt.

timwoolliscroft-opteran commented:
> Change as itself looks fine. But then the if/else section in between should then be modified. This should be done with #55987 in mind. Also I might be wrong but one bug remain about the secondary slave matching in case address is 10bits. This is not related to your change, but if you have the set up to check this, it would be great if you can confirm my doubt.

Yes secondary slave matching with 10bit addressing is broken by the looks of it, this fix doesn't address #55987 in any way.

I'm not sure about the if/else section - I think its valid as its the safest way to catch the no match condition and ASSERT.

Sorry I'm not really the right person to take on #55987, this is my first STM32 in 8 years & first zephyrproject (started in anger on Monday).

timwoolliscroft-opteran commented:
Seems like this bug was introduced in this PR...
https://github.com/zephyrproject-rtos/zephyr/pull/60620

Reading comments around the HAL for my device...
https://github.com/STMicroelectronics/stm32h7xx_hal_driver/blob/dca6b63d9cce505f2801107f514c491d850a6cbb/Inc/stm32h7xx_ll_i2c.h#L2171

It seems ST might not fully support dual 10 addressing... Looking at Tables 387 in the I2C_ISR register only has 7bits of ADDRCODE, see:
 
https://www.st.com/resource/en/reference_manual/rm0455-stm32h7a37b3-and-stm32h7b0-value-line-advanced-armbased-32bit-mcus-stmicroelectronics.pdf

erwango commented:
> Also interesting in that table I2C_OAR1.OA is 10bit - where as I2C_OAR2.OA is 7bit. So its never going to work nicely.

Good catch, so that would explain implementation proposed in https://github.com/zephyrproject-rtos/zephyr/pull/60620 (on this part at least)

> I would suggest the Zephyr STM32 i2c driver should error if a secondary target address is attempted in 10 bit mode.

Yep.

timwoolliscroft-opteran commented:
@erwango - could you advise best direction forward? Do you require further changes? 

For me the fix preserves existing functionality of the 10bit mode whilst also fixing the actual bug ticket. Dual addressing in 10 bit mode seems really a bit out of scope. And I'm not sure we have considered fully what current users might be thinking they are getting.

erwango commented:
> Do you require further changes?

At least removing the redundant "if" part would be nice, if it isn't too much to ask.

timwoolliscroft-opteran commented:
The `if` is not redundant. It is forming part of the chain that leads to assert. I removed and tested and got the assert.

I've reworked a little to make it more readable but also protect the 10 bit mode from a NULL slave_cfg.

erwango reviewed (commented):
Thanks for going the extra mile.

One requirement is that you shouldn't introduce commits fixing initial PR commits. So in current case, commits should all be squashed.
Reason is that when merging, we don't create a marge commit: pull request is rebased and merged "as is", on top of main branch. So for a given fix, one commit is expected.
(Several commits would be expected when change impacts several areas or when change is complex and then splitting in smaller commits helps readability and speeds up review process)

timwoolliscroft-opteran committed:
drivers: i2c: stm32: Fix routing of secondary target address

Fixes the routing of the events associated with a secondary
i2c target address being routed to the primary config. The
i2c_target_config/slave_cfg was being selected from the
driver address match but then over written by the primary.
This change fully implements the if/else of 10bit addressing
and includes a assert if the slave_cfg is NULL, and explains
why dual 10bit addresses on STM32 won't work.

Signed-off-by: Tim Woolliscroft <tim@opteran.com>

timwoolliscroft-opteran commented:
Learning more and more git everyday! Be honest internally we just put a merge commit in so not used to other peoples rules and regulations. Added to that I'm historically a hardware engineer not a software developer.

Hope we're all good now!

erwango commented:
> Learning more and more git everyday! Be honest internally we just put a merge commit in so not used to other peoples rules and regulations. Added to that I'm historically a hardware engineer not a software developer.

No problem. Note also that each project has its own rules, so this is not obvious.

erwango reviewed (approved):
Looks all fine. Thanks for the changes!
